New implementation of Knutson scaling incl. fix to negative freqs and double counting

aleeciu opened the pull request:
In this PR I implement new scaling factors to simulate TC under future climate. The new implementation follows the same logic as the previous one but here new scaling factors are used. The new scaling factors are based on [Knutson et al.(2020)](https://journals.ametsoc.org/view/journals/bams/101/3/bams-d-18-0194.1.xml) and [Jewson et al. (2021)](https://journals.ametsoc.org/view/journals/apme/60/11/JAMC-D-21-0102.1.xml). 

Important to note is that:

- following [Jewson (2022)](https://link.springer.com/article/10.1007/s00477-021-02142-6), only frequencies are scaled in order to avoid inconsistencies and possible double-counting as also raised in #453. Scaling only frequencies anyway changes mean intensities too. 
- in order to avoid any overlap in frequency changes across categories, frequencies for categories 0-3 are derived from those of categories 0-5 and 4-5, and all TCs in categories 0-3 or 4-5 are then corrected with their respective factors.
- this PR also fixes the occurrence of negative frequencies reported in #317 when using the old scaling factors.

I created a [jupyter notebook](https://drive.google.com/drive/u/0/folders/1AFVmdzOmYEmJPY6nC9Z2sE0r6vepI9le) where I show how the results of such implementation lead to outputs which are consistent with the estimates from Knutson 2020 and never lead to negative frequencies. That said, since the test is done under a given baseline period, I cannot rule out the possibility that negative frequencies will never occurr, therefore I am anyway throwing an error when this occur.

Two remaining todos:
- final tests, as I first want to make sure that the followed approached is accepted
- the current TC tutorial shows an increase in absolute intensity of Irma under a given RCP. However, as the new implementation does not apply any increment in absolute intensity but, rather, to mean intensities and any other risk metric, I was wondering how to best rewrite that part.
- possibly, the notebook I posted above could potentially be simplfied and become and TC_CC tutorial, which is currently missing

### PR Author Checklist

- [ ] Read the [Contribution Guide][contrib]
- [x] Correct target branch selected (if unsure, select `develop`)
- [x] Descriptive pull request title added
- [] Source branch up-to-date with target branch
- [] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [] [Tests][testing] updated
- [] [Tests][testing] passing
- [ ] No new [linter issues][linter]
- [ ] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Reviewer_Checklist.html) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html
[linter]: https://climada-python.readthedocs.io/en/stable/guide/Guide_Continuous_Integration_and_Testing.html#3.C.--Static-Code-Analysis

chahank commented:
Thank you for the proposal. Note that we have already had more than one trial to improve on this. So, here are two questions (which you might have solved):

- How do you deal with the significance of the Knutson parameters? 
- How do you incorporate the fact that the Knutson analysis was done on completely different datasets? 
- If you just remove the intensity scaling, you do not replicate the distribution shift describe in the paper by Knutson. How/why is it clearly better to have only the frequency shifting?
- How do you fix the negative frequencies without messing up the distribution shift?

bguillod commented:
Thanks for tackling the issues related to future TC projections.

I personally think this implementation represents a substantial improvements compared to the current code. It is not perfect, but this is to a wide extent due to the nature of the complex problem and I think the proposed solution is a good pragmatic way to solve the main issues. The implementation as you did here is in my view very much in line with the literature.

One concern I would have, however, is related to how Knutson presents changes and then how you implemented them. Knutson gives changes for "intensity above cat X", so we have changes for categories (ordered from highest to lowest category range): 

1. 4 to 5
2. 3 to 5
3. 1 to 5
4. 0 (i.e. tropical storms) to 5

So for example storms of category 3 fall under both criteria (1) and (2) above. Which change should one apply?
My suggestion would be:
- Calculate the frequency of events by category and basin in the input dataset (i.e. how many storms in your event set by category and basin).
- For events where only one frequency change is available from Knutson (i.e., categories 4 and 5 in the example above as only point 1 applies), apply that change.
- Then go through critera 2,3,4 and each time, look at how frequency has already changed and how to adjust remaining categories such as to satisfy Knutson's aggregated change. That is, for point 2, I would look at cat 3 storms and see: by how much do these need to change in frequency after we've already changed cat 4 and 5 storms, such as to satisfy Knutson's frequency change (1) and (2) above. Once this is done, do the same for cat 1&2 (Knutson's change 3 above), etc.

That way you would be able to conserve the changes for all category ranges that Knutson is providing.

If you can solve that then I think this will bring the implementation to a new level.

chahank commented:
> Then go through critera 2,3,4 and each time, look at how frequency has already changed and how to adjust remaining categories such as to satisfy Knutson's aggregated change. That is, for point 2, I would look at cat 3 storms and see: by how much do these need to change in frequency after we've already changed cat 4 and 5 storms, such as to satisfy Knutson's frequency change (1) and (2) above. Once this is done, do the same for cat 1&2 (Knutson's change 3 above), etc.

Exactly this method was implemented in the past. But, it did lead to strong inconsistencies with the frequency distributions. A new implementation should be aware of this. 

This was the proposal of the time. Maybe only changing the frequencies would resolve the problem.

```
 def _apply_knutson_criterion(self, chg_int_freq, scaling_rcp_year):
        """
        Apply changes to intensities and cumulative frequencies.

        Parameters
        ----------
        criterion : list(dict))
            list of criteria from climada.hazard.tc_clim_change
        scale : float
            scale parameter because of chosen year and RCP
        Returns
        -------
        tc_cc : climada.hazard.TropCyclone
            Tropical cyclone with frequency and intensity scaled according
            to the Knutson criterion. Returns a new instance of TropCyclone.
        """

        tc_cc = copy.deepcopy(self)

        # Criterion per basin
        for basin in np.unique(tc_cc.basin):

            bas_sel = (np.array(tc_cc.basin) == basin)

            # Apply intensity change
            inten_chg = [chg
                         for chg in chg_int_freq
                         if (chg['variable'] == 'intensity' and
                             chg['basin'] == basin)
                         ]
            for chg in inten_chg:
                sel_cat_chg = np.isin(tc_cc.category, chg['category']) & bas_sel
                inten_scaling = 1 + (chg['change'] - 1) * scaling_rcp_year
                tc_cc.intensity = sparse.diags(
                    np.where(sel_cat_chg, inten_scaling, 1)
                    ).dot(tc_cc.intensity)

            # Apply frequency change
            freq_chg = [chg
                        for chg in chg_int_freq
                        if (chg['variable'] == 'frequency' and
                            chg['basin'] == basin)
                        ]
            freq_chg.sort(reverse=False, key=lambda x: len(x['category']))

            # Iteratively scale frequencies for each category such that
            # cumulative frequencies are scaled according to Knutson criterion.

            cat_larger_list = []
            for chg in freq_chg:
                cat_chg_list = [cat
                                for cat in chg['category']
                                if cat not in cat_larger_list
                                ]
                sel_cat_chg = np.isin(tc_cc.category, cat_chg_list) & bas_sel
                if sel_cat_chg.any():
                    freq_scaling = 1 + (chg['change'] - 1) * scaling_rcp_year
                    sel_cat_all = (np.isin(tc_cc.category, chg['category'])
                                   & bas_sel)
                    sel_cat_larger = (np.isin(tc_cc.category, cat_larger_list)
                                      & bas_sel)
                    freq_scaling_cor = (
                        (np.sum(self.frequency[sel_cat_all]) * freq_scaling
                         - np.sum(tc_cc.frequency[sel_cat_larger]))
                        / np.sum(self.frequency[sel_cat_chg])
                    )
                    tc_cc.frequency[sel_cat_chg] *= freq_scaling_cor
                cat_larger_list += cat_chg_list

        if (tc_cc.frequency < 0).any():
            raise ValueError("The application of the given climate scenario"
                             "resulted in at least one negative frequenciy."
                             "This is likely due to the use of a"
                             "non-representative event set (too small, "
                             "incorrect reference period, ...)")

        return tc_cc
```

bguillod commented:
> Exactly this method was implemented in the past. But, it did lead to strong inconsistencies with the frequency distributions. A new implementation should be aware of this.

I agree with you @chahank that this back-then-proposed implementation you've copy-pasted above would do what I meant for frequencies. Can you please elaborate on what the inconsistencies were in the distribution? Is this discussed somewhere in an issue or an open PR or something?

> This was the proposal of the time. Maybe only changing the frequencies would resolve the problem.

It might be worth a try. Surely, applying both changing in frequency and intensity must mess up the distribution, but this does not mean that doing only frequency wouldn't. So first I'd like to understand better the inconsistencies you mentioned.

aleeciu commented:
Hi both,

to me both the proposal by @bguillod and the code posted by @chahank make sense, but happy to hear what inconsistencies you encountered Chahan.

I am just not fully sure about this @bguillod :

> Calculate the frequency of events by category and basin in the input dataset (i.e. how many storms in your event set by category and basin).

are you suggesting to calc yet other - cat-specific - frequencies other than those carried by the TropCyclone object which account for *all* events - regardless their cat - in a basin ?

bguillod commented:
> I am just not fully sure about this @bguillod :
> 
> > Calculate the frequency of events by category and basin in the input dataset (i.e. how many storms in your event set by category and basin).
> 
> are you suggesting to calc yet other - cat-specific - frequencies other than those carried by the TropCyclone object which account for _all_ events - regardless their cat - in a basin ?

No, what I meant is essentially what is implement in the code shared by @chahank, I.e. summing up all frequencies of events by basin and category ranges.

On a side-note: find the code extremely hard to understand though. We could definitely improve readability.

bguillod commented:
Do you remember for which basin and scenario & year this was the case?
Perhaps @aleeciu could use historical data, apply changes in frequencies only using this methodology and see if and for which cases negative frequencies occur?

chahank commented:
Unfortunately no, it was quite some time ago. The cleanest would be imho to redo the analysis with the latest version of the random walk algorithm.

chahank commented on climada/hazard/trop_cyclone.py:
```diff
@@ -629,16 +629,39 @@ def _apply_knutson_criterion(
+                        / np.sum(self.frequency[sel_cat_chg])
+                    )
+
+                    if freq_scaling_cor < 0:
+                        LOGGER.warning("The application of the given climate scenario"
+                                       " would result in negative frequency correction factors."
+                                       " Frequencies are thus left unchanged but this"
+                                       " should warrant attention from the user.")
+                        freq_scaling_cor = 1
```
I am not sure that this is a good handling of this case. A user will simply not know what to do here. So, we should either throw an error or find an actual solution that is defendable. Is just not scaling the frequencies really a good solution?

chahank commented:
Is it still planed to finish this pull request?

aleeciu commented:
> Is it still planed to finish this pull request?

Yes, but I need to figure out how to address the points you raised above.

aleeciu commented:
Ok, I completely changed the way I address the fixes in this PR. I changed the whole description accordingly, and provide detailed info of the new approach incl. a link to a jupyter notebook that showcases results obtained with the new scaling.

bguillod commented on climada/hazard/tc_clim_change.py:
```diff
@@ -1,174 +1,329 @@
 
-You should have received a copy of the GNU General Public License along
-with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+def get_knutson_scaling_factor(
```
Is my understanding correct that this function returns the scaling factors every 5 years from 2000 to 2100?
Why not rather return only the factors for one target year here?

bguillod commented on climada/hazard/tc_clim_change.py:
```diff
@@ -1,174 +1,329 @@
+    return pd.DataFrame(predicted_property_pcs,
+                        index=mid_years, columns=rcps)
+
+def get_gmst_info():
```
Would it be clearer to the user if this function rather returned a pd.DataFrame with columns 'year', 'rcp', 'gsmt'? Or would this make the calculations more cumbersome later on?

bguillod commented on climada/hazard/tc_clim_change.py:
```diff
@@ -1,174 +1,329 @@
+        response to anthropogenic warming. Bull. Amer. Meteor. Soc., 101 (3), E303–E322, 
+        https://doi.org/10.1175/BAMS-D-18-0194.1.
+    
+    for 4 variables (i.e., cat05 frequency, cat45 frequency, intensity precipitation rate), 
```
I think you actually mean 4 variables, but you only list 3 here.

bguillod commented on climada/hazard/tc_clim_change.py:
```diff
@@ -1,174 +1,329 @@
+
+    knutson_data = np.empty((4, 6, 5))
+
+    # fig 1 in Knutson et al. 2020 (except min, max metrics and global region)
```
Can you in this and the similar comments below also specify which variable these data are for?

bguillod commented on climada/hazard/trop_cyclone.py:
```diff
@@ -336,46 +336,93 @@ def from_tracks(
         self,
-        ref_year: int = 2050,
-        rcp_scenario: int = 45
+        pct: int = 2,
```
In my opinion:

1. Any user calling this function should explicitly specify for which future year and rcp scenario they are targetting. Most likely, the baseline years is also something you would like to force the user to provide, although I can see why a default could also be considered for that argument. Input parameter `pct`, that's for sure, has a clear default: the median.
2. Rather than using an argument named `pct` which is actually an encoded value, I suggest naming it `percentile` and its value to be the actual percentile (default: 50), with a quick check at the beginning that it is one of the allowed values (if not, raise an error). Or is the encoding because value 0 refer sometimes to 5 or 10 and value 4 sometimes to 90 or 95? In that case, can we make this consistent by using 10/90 and explain in the docstring that the 10/90 actually refer to 5/95 for the cat 0-5 frequencies?

Therefore I suggest to use the following function definition:
```
    def apply_climate_scenario_knu(
        self,
        ref_year: int,
        rcp_scenario: int,
        baseline: tuple,
        percentile: int = 50,
    ):
```
If you go for this, the docstring will have to be adjusted accordingly (and unit tests as well I presume).

bguillod replied:
Alternatively, if `pct` (which I'd rename anyway) cannot be 10/25/50 etc, then perhaps we could use a string encoding such as 'median', 'IQR-', 'IQR+', 'high', 'low'. Just a spontaneous idea here.

bguillod commented on climada/hazard/trop_cyclone.py:
```diff
@@ -336,46 +336,93 @@ def from_tracks(
+        baseline : tuple of int
+            the starting and ending years that define the historical baseline.
+        pct: int
+            percentile of interest:
```
Be more specific: this percentile refers to the percentile of all models' individual projections? Maybe something like "percentile of model estimates"?

bguillod reviewed (changes requested):
Thanks @aleeciu for these improvements! Science-wise I think they are a clear improvement and remove many of the shortcomings of the previous methodology, therefore I think these changes should be merged.
Before merging, I have a few comments in terms of code. Some of the comments I made in `trop_cyclone.py` might imply adjustments to `tc_clim_change.py` and vice-versa; try to make both file (e.g. function APIs) as consistent as possible as you go through them.

I have not yet gone through the unit tests but will be happy to do so once you have addressed this first set of comments.

bguillod commented on climada/hazard/trop_cyclone.py:
```diff
@@ -336,46 +336,93 @@ def from_tracks(
+        percentile: str = '50',
+        rcp_scenario: int = 85,
+        ref_year: int = 2035,
+        baseline: tuple = (1982, 2022)
```
Personally and as commented on the previous version, I would not set any default values for all the parameters, in order to force the user to select a scenario, percentile, year and baseline. Also, I am not sure about the name `ref_year`, why not `target_year` instead? Seems more implicit to me.
Finally, I would change the order of the arguments, using in the order: target_year, rcp_scenario, percentile, baseline.

bguillod commented on climada/hazard/trop_cyclone.py:
```diff
@@ -336,46 +336,93 @@ def from_tracks(
         Parameters
         ----------
         ref_year : int
             year between 2000 ad 2100. Default: 2050
```
The description of `ref_year` is not very clear. Perhaps "future year to be simulated, between 2000 and 2100."?

bguillod commented on climada/hazard/trop_cyclone.py:
```diff
@@ -336,46 +336,93 @@ def from_tracks(
+        frequency and intensity changes, as the relatioship between these two is non
+        trivial to resolve.
 
         Parameters
```
I would describe the parameters in the same order as they appear in the function call.

bguillod commented on climada/hazard/trop_cyclone.py:
```diff
@@ -336,46 +336,93 @@ def from_tracks(
+        baseline : tuple of int
+            the starting and ending years that define the historical baseline.
+        percentile: str
+            percentile of Knutson et al. 2020 estimates. These estimates come from a
```
I suggest "percentile of Knutson et al. 2020 estimates, representing the model uncertainty in future changes in TC activity" for the first sentence.

bguillod reviewed (commented):
Thanks for the adjustments! I have 4 small remaining comments and after that it will be ready to merge from my perspective.

aleeciu commented:
> Thanks for the adjustments! I have 4 small remaining comments and after that it will be ready to merge from my perspective.

thank @bguillod - I implemented the suggested changes.

bguillod commented:
From my point of view this is ready to merge - leaving it up to @chahank or whoever to review and/or merge as appropriate.

chahank commented:
Ok, thanks! I will review this timely.

chahank commented on climada/hazard/tc_clim_change.py:
```diff
@@ -1,174 +1,375 @@
-    rcp_scenario: int, optional
-        26 for RCP 2.6, 45 for RCP 4.5. The default is 45
-        60 for RCP 6.0 and 85 for RCP 8.5.
+    Tropical cyclones and climate change assessment. Part II: Projected 
```
Great to add the sources! Could you please clean-up the formatting?

peanutfun replied:
Please resolve all `trailing-whitespace` issues encountered by Jenkins. You see the respective annotations in the "Changes" tab of the PR

chahank commented on climada/hazard/tc_clim_change.py:
```diff
@@ -1,174 +1,375 @@
+    knutson_data : np.array of dimension (4x6x5)
+        array contaning data used by Knutson et al. (2020) to project changes in cat05 frequency,
+        cat45 frequency, intensity and precipitation rate (first array's dimension), for the
+        N. Atl, NW Pac., NE Pac., N. Ind, S. Ind., SW Pac. regions (second array's dimension)
```
These are different basin names as in the rest of the code. Why? What is the relation?

chahank commented on climada/hazard/tc_clim_change.py:
```diff
@@ -1,174 +1,375 @@
-    factor : float
-        factor to scale Knutson parameters to the give RCP and year
+    knutson_data : np.array of dimension (4x6x5)
+        array contaning data used by Knutson et al. (2020) to project changes in cat05 frequency,
+        cat45 frequency, intensity and precipitation rate (first array's dimension), for the
```
What are `cat05` and `cat45` ?

What do precipitation rates have to do here? Are they handled in this code?

chahank commented on climada/hazard/tc_clim_change.py:
```diff
@@ -1,174 +1,375 @@
+
+    # fig 1 in Knutson et al. 2020
+    # (except min, max metrics and global region): cat05 frequency 
+    knutson_data[0,0]=[-34.49,-24.875,-14.444,3.019,28.737]
```
Are these the data from the figure?

Then again I would define them as default parameters instead of hiding them in a function.

peanutfun replied:
To re-iterate @chahank's point: It is impossible to read these exact values from the figure. How were they extracted? The figure in question:

<img width="792" alt="Screenshot 2024-05-06 at 11 22 59" src="https://github.com/CLIMADA-project/climada_python/assets/34276446/06f0ec5c-5b57-4a4b-b3ce-299431f2193a">

It might also happen to structure the data differently, e.g. with a DataFrame and named columns ("5th percentile"), so it's always clear which array dimension means what.

chahank commented on climada/hazard/test/test_tc_cc.py:
```diff
@@ -26,40 +26,51 @@
+        self.assertAlmostEqual(criterion.loc[2070, '6.0'], -31.06633, 4)
+        self.assertAlmostEqual(criterion.loc[2100, '8.5'], -58.98637, 4)
+
+    def test_get_gmst_pass(self):
```
Could you please write this as a simple unit test (it must be plainly obvious when the answer is correct or not). Using more complex / real data is better for an integration test.

peanutfun replied:
New/updated tests should not follow outdated test logic. Anyway, all testing of `get_gmst_info` is somewhat pointless. The function takes no arguments and only returns structured data. You should instead test if the data is correctly applied in other functions calling `get_gmst_info`

chahank commented on climada/hazard/test/test_tc_cc.py:
```diff
@@ -26,40 +26,51 @@
+        self.assertAlmostEqual(gmst_info['gmst_data'][-1,0], -0.16)
+        self.assertAlmostEqual(gmst_info['gmst_data'][-1,-1], 4.477764, 4)
+
+    def test_get_knutson_data_pass(self):
```
Could you please write this as a simple unit test (it must be plainly obvious when the answer is correct or not). Using more complex / real data is better for an integration test.

chahank commented on climada/hazard/test/test_trop_cyclone.py:
```diff
@@ -413,29 +414,33 @@ def test_apply_criterion_track(self):
-        self.assertTrue(np.allclose(tc.intensity[1, :].toarray(), tc_cc.intensity[1, :].toarray()))
-        self.assertTrue(np.allclose(tc.intensity[3, :].toarray(), tc_cc.intensity[3, :].toarray()))
+        tc_cc = tc.apply_climate_scenario_knu(percentile='50',
+                                              rcp_scenario=45,
+                                              target_year=2050)
+        self.assertFalse(
+            np.allclose(tc.frequency[1], tc_cc.frequency[1])
+            )
+        self.assertFalse(
+            np.allclose(tc.frequency[0], tc_cc.frequency[0])
+        )
         self.assertFalse(
-            np.allclose(tc.intensity[0, :].toarray(), tc_cc.intensity[0, :].toarray()))
+            np.allclose(tc.frequency[2], tc_cc.frequency[2])
+            )
```
This appears to be a very weak test where one only checks that the values changed, but not that the values are correct. Is this intended?

chahank commented on climada/hazard/test/test_trop_cyclone.py:
```diff
@@ -444,100 +449,67 @@ def test_apply_criterion_track(self):
-            basin=['SP', 'SP'],
-            category=np.array([0, 1]),
-        )
+        NA_scaling_03 = (NA_scaling_05 * np.sum(tc.frequency[cat05_sel & NA_bas_sel]) 
+                         - NA_scaling_45 * np.sum(tc.frequency[cat45_sel & NA_bas_sel])
+                        ) / np.sum(tc.frequency[cat03_sel & NA_bas_sel])
```
How can one know that this is the correct answer? Is this obvious?

chahank commented on climada/hazard/trop_cyclone.py:
```diff
@@ -336,46 +336,93 @@ def from_tracks(
+        percentile: str
+            percentile of Knutson et al. 2020 estimates, representing the model
+            uncertainty in future changes in TC activity. These estimates come from a
+            review of state-of-the-art literature and models. As extreme percentiles,
+            the 5th and 95th are reported for variable cat05 and the 10th and 90th
+            for cat45 and intensity. All three (cat05, cat45, intensity) report the
+            25th, 50th and 75th percentiles. Refer to the mentioned publications above
+            for more details. Possible values are:
+                '5/10', '25', '50', '75', '90/95'
```
Can you please add a recommendation for when to use what? This is a rather obscure explanation, and the names `cat05`, `cat45`, `intensity` are unclear here.

chahank commented on climada/hazard/trop_cyclone.py:
```diff
@@ -336,46 +336,93 @@ def from_tracks(
+                85 for RCP 8.5
+        target_year : int
+            future year to be simulated, between 2000 and 2100.
+        baseline : tuple of int
+            the starting and ending years that define the historical baseline.
```
Default is missing. Any reason for the choice of years?

chahank commented on climada/hazard/trop_cyclone.py:
```diff
@@ -336,46 +336,93 @@ def from_tracks(
-        haz_cc = self._apply_knutson_criterion(chg_int_freq, scale_rcp_year)
-        return haz_cc
+
+        map_rcp_names = {26: '2.6', 45: '4.5', 60: '6.0', 85: '8.5'}
```
Why not use these values directly as input?

chahank commented on climada/hazard/trop_cyclone.py:
```diff
@@ -336,46 +336,93 @@ def from_tracks(
+
+            bas_sel = np.array(tc_cc.basin) == basin
+
+            scale_year_rcp_03 = (scale_year_rcp_05 * np.sum(tc_cc.frequency[sel_cat05 & bas_sel]) -
+                                scale_year_rcp_45 * np.sum(tc_cc.frequency[sel_cat45 & bas_sel])
+                                ) / np.sum(tc_cc.frequency[sel_cat03 & bas_sel])
```
Please make a function out of this that can be properly explained and tested.

chahank reviewed (changes requested):
This is a great improvement and will make CLIMADA able to deliver much better TC climate change estimates. Thank you very much for this contribution!

I see some room for improvement in particular in the code readability. Also, it would be very good to have a document with DOI (can be very brief) which explains the here implemented methods. For instance, what are the input TC datasets that would be suitable here? What are the limitations?

aleeciu commented:
Many thanks @chahank for the extensive review. I will follow on it between this and next week.

Just one thing to clarify: The idea of this PR was to integrate a recently published method - and related published code - into CLIMADA to improve the TC scaling. This implies that the code at https://github.com/CLIMADA-project/climada_python/blob/23e7404f54409c9b264063357cb6daa97a52a27e/climada/hazard/tc_clim_change.py is mostly in its original form with only slight adaptations from my side, which is also why that file gets a different license agreement that the rest of CLIMADA (we discussed it on slack). Now, many of your questions point to that very file, but addressing them would require me to change it quite a bit, which would be an overkill since I would not end up integrating existing code but rather reimplement the whole methodology.

I don't know what's best here.

chahank commented:
> Many thanks @chahank for the extensive review. I will follow on it between this and next week.
Great, thanks. 

> Just one thing to clarify: The idea of this PR was to integrate a recently published method - and related published code - > into CLIMADA to improve the TC scaling. This implies that the code at https://github.com/CLIMADA-project/climada_python/blob/23e7404f54409c9b264063357cb6daa97a52a27e/climada/hazard/tc_clim_change.py is mostly in its original form with only slight adaptations from my side, which is also why that file gets a different license agreement than the rest of CLIMADA (we discussed it on slack). Now, many of your questions point to that very file, but addressing them would require me to change it quite a bit, which would be an overkill since I would not end up integrating existing code but rather reimplementing the whole methodology.

As you know, we are committed to producing high-quality, stable and maintainable code with CLIMADA. The changes proposed here are required to fulfil these goals. I also think that overall the changes are rather minor (some refactoring of the methods to be clearer). 

> I don't know what's best here.

peanutfun commented on climada/hazard/tc_clim_change.py:
```diff
@@ -1,174 +1,375 @@
-"""
-This file is part of CLIMADA.
+#
+# MIT License
+#
+# Copyright (c) 2021 Lynne Jewson, Stephen Jewson
+# Copyright (c) 2023 ETH Zurich, CLIMADA contributors listed in AUTHORS
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
```
If this code is based on MIT-licensed code (@aleeciu could you please clarify?), then we should be able to incorporate a modified version into our GPL-licensed software and distribute it under GPL (only). However, we would have to retain this notice and the copyright statement in particular. I am not entirely sure about these things, and we should probably discuss that with the ETH legal department.

peanutfun replied:
Discussed today: The MIT license is GPL-compatible. @aleeciu Please remove this comment and instead place the Climada module docstring at the top of the file like so:
```python
"""
This file is part of CLIMADA.

Copyright (C) 2017 ETH Zurich, CLIMADA contributors listed in AUTHORS.

CLIMADA is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free
Software Foundation, version 3.

Parts of this module are modifications of work originally published under the following license:

MIT License

Copyright (c) 2021 Lynne Jewson, Stephen Jewson

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

---

<module description here>
"""

peanutfun commented:
@aleeciu This PR has been stale for three months. @chahank requested several changes in the review. Can you estimate when you will update the code accordingly?

We will pick up the question about licensing in the next Senate meeting. Sorry for the delay.

peanutfun commented:
@aleeciu What's the status?

Licensing issue has been clarified, please see https://github.com/CLIMADA-project/climada_python/pull/734#discussion_r1577974039

aleeciu commented:
Hi @chahank , @peanutfun , @bguillod,
many thanks for the inputs and sorry for the delay, but I guess 3 months to revise it's still ok if it takes 1+ month to review :-))) ? 

I implemented most suggested changes, in particular toward improving code readability, docstrings, variables descriptions etc. as well as I added the license statement. Left are comments regarding tests, which I personally find out of scope for this PR, as the ones I wrote follow the same logic and structure as the already existing ones. I am alinged we might need better tests across modules, but then this is the scope of a dedicated PR.

Cheers, Alessio

chahank commented:
> Hi @chahank , @peanutfun , @bguillod, many thanks for the inputs and sorry for the delay, but I guess 3 months to revise it's still ok if it takes 1+ month to review :-))) ?
> 
> I implemented most suggested changes, in particular those towards improving code readability, docstrings, variables descriptions etc., as well as I added the license statement. Left are comments regarding tests, which I personally find out of scope for this PR, as the ones I wrote follow the same logic and structure as the already existing ones. I am alinged we might need better tests across modules, but then this is the scope of a dedicated PR.
> 
> Also, current tests crash although they run just fine on my machine. I am not too sure but it seems it has to do with the `tc_rainfield` module which is however not touched in the PR.
> 
> Cheers, Alessio

Thanks for the update! Excellent work. 

Regarding the tests, please recall that we have clear contribution guidelines for tests (which have been in place for a long time, so you should be aware of them). In there, it is clear that copying very old tests is not the way forward. It would have been faster to immediately write the tests in line with the guidelines. I hope this will happen for future pull requests from your team. 

Now, for the current PR, if @peanutfun agrees, I am ok with leaving them as is and tackling tests in potential future PR.

For the TC-rain test crash, this is indeed unrelated to your PR, so please ignore it :D.

chahank commented:
Two points:

- could you please make sure to remove trailing white spaces (this can be configured in your code editor in general), otherwise the linter complains.
- the tests are now failing due to an error in the newly implemented code. This needs to fixed.

aleeciu commented:
> Two points:
> 
> * could you please make sure to remove trailing white spaces (this can be configured in your code editor in general), otherwise the linter complains.
> * the tests are now failing due to an error in the newly implemented code. This needs to fixed.

Thanks:
- I have removed those my editor found, or is it complaining also on other stuff?
- I fixed the tests. Forgot to update TropCyclones test

Also, I realized it's much cleaner to infer baseline from the event set dates, and so the baseline argument is no longer required.

peanutfun commented on climada/hazard/test/test_trop_cyclone.py:
```diff
@@ -503,29 +504,34 @@ def test_apply_criterion_track(self):
+        self.assertFalse(
+            np.allclose(tc.frequency[1], tc_cc.frequency[1])
+            )
         self.assertFalse(
-            np.allclose(tc.intensity[0, :].toarray(), tc_cc.intensity[0, :].toarray()))
+            np.allclose(tc.frequency[0], tc_cc.frequency[0])
+        )
         self.assertFalse(
-            np.allclose(tc.intensity[2, :].toarray(), tc_cc.intensity[2, :].toarray()))
-        self.assertTrue(np.allclose(tc.frequency, tc_cc.frequency))
+            np.allclose(tc.frequency[2], tc_cc.frequency[2])
+            )
+        self.assertFalse(
+            np.allclose(tc.frequency[3], tc_cc.frequency[3])
+            )
```
These tests check that the frequencies are ... not the same? Why are intensities not checked anymore? What does the "not the same" check achieve? These tests should check the actual results. 

Please always use `np.testing.assert_allclose(x, y)` instead of `self.assertTrue(np.allclose(x, y))` to get a useful error message when the test fails.

aleeciu replied:
Yes I am testing that the module actually have changed the frequencies, which is essence all it does. Intensities are _directly_ changed anymore, because that would double counting. The test after this one actually tests that the operations that change these frequencies work.

Could you please show me how to use `np.testing.assert_allclose(x, y)` in the negative case? simply with a tilde?

peanutfun replied:
Your tests should not check that the values have changed, but should instead check the resulting values. For that reason, there is no negative case of `np.testing.assert_allclose`. Note that such a negative case would even succeed of one of the arrays suddenly turned into NaNs.

peanutfun reviewed (changes requested):
@aleeciu I only asked for a status update. There is no problem with stating that you need more time! We would only take the liberty to close PRs if we get no responses from authors for a longer period.

I generally agree with @chahank that it's inacceptable to add new tests that don't follow our guidelines only because the existing tests for the same module don't either. Writing guideline-compliant tests for new code is clearly in the scope of every PR.

That being said, I am fine with keeping most of this PR as is and merging it, like @chahank. However, please fix the tests in `test_apply_climate_scenario_track`, which only seem to check that TC frequencies have changed somehow after the climate scenario was applied. See my comment.

aleeciu commented:
> @aleeciu I only asked for a status update. There is no problem with stating that you need more time! We would only take the liberty to close PRs if we get no responses from authors for a longer period.
> 
> I generally agree with @chahank that it's inacceptable to add new tests that don't follow our guidelines only because the existing tests for the same module don't either. Writing guideline-compliant tests for new code is clearly in the scope of every PR.
> 
> That being said, I am fine with keeping most of this PR as is and merging it, like @chahank. However, please fix the tests in `test_apply_climate_scenario_track`, which only seem to check that TC frequencies have changed somehow after the climate scenario was applied. See my comment.

Thanks a lot and I indeed get the point about tests: it's just that - very simply put - for somewhat irregular contributors like me with only decent coding skills it's an non-negligible overhead to come up with brand new tests. So we would either need support from the core team to do this or let it be and address it differently (via hackatons, climada days etc.). The price would be that these contributions would not get through at all, but that's a pity because e.g. in this case this is a major bug fix.

chahank commented:
> Thanks a lot and I indeed get the point about tests: it's just that - very simply put - for somewhat irregular contributors like me with only decent coding skills it's an non-negligible overhead to come up with brand new tests. So we would either need support from the core team to do this or let it be and address it differently (via hackatons, climada days etc.). If not, the price would be that these contributions would not get through at all, but that's a pity because e.g. in this case this is a major bug fix.

I am quite confused by this. You have been working on CLIMADA for almost 3 years and were present at the CLIMADA days when we demonstrated how to write unit tests. Besides, following the guide and learning basic unit test writing requires much less time than what we are spending now on the review. Writing good tests is NOT optional. Not following the carefully written guidelines, which balance the need for code quality and the required working time, is just creating unnecessary work for the core reviewer team and in addition is likely to lead to technical dept.

aleeciu commented:
Hey, I really don't understand why you are getting off the roof here or why you are making it so personal. I just made a general statement about contributors' groups and I don't know how saying that there is a core (three or four ppl) and an irregular (everyone else) contributing group can surprise you to be honest.

Indeed I attended the days, but that does make me a pro. Also, for the sake of logic, we shall then shut down all _existing_ modules whose tests are not complaint, or not? 
Anyway, let me cut this short: I will make these unit tests. In terms of timeline, I will probably find the time in the next couple of months.

Take care, Alessio

aleeciu commented:
Understood, all good. Let me fixed this thing about tests and meanwhile please let me know if there are feedback on the rest.

peanutfun commented on climada/hazard/tc_clim_change.py:
```diff
@@ -1,174 +1,375 @@
-        {'basin': 'SP', 'category': [4, 5],
-         'year': 2100, 'change': 1 - 0.583, 'variable': 'frequency'}
-        ]
+    gmst_info.update({'gmst_data':
+        np.empty((len(gmst_info['rcps']),
+                  gmst_info['gmst_end_year']-gmst_info['gmst_start_year']+1))
+    })
```
You are setting an empty array here but then set the entries in the array in the next step. Instead, you could simply initialize the array with the values directly. Compare:
```python
# What you essentially do
arr = np.empty((2, 2))
arr[0] = [0, 1]
arr[1] = [2, 3]

# What you could do
arr = np.array([
   [0, 1],
   [2, 3],
])
```

peanutfun commented on climada/hazard/tc_clim_change.py:
```diff
@@ -1,174 +1,375 @@
-        {'basin': 'NI', 'category': [4, 5],
-         'year': 2100, 'change': 1, 'variable': 'frequency'}
-        ]
+    gmst_info['gmst_data'][0] = [
```
Then please add in the docstring of this method where the data comes from.

peanutfun commented on climada/hazard/tc_clim_change.py:
```diff
@@ -1,174 +1,375 @@
-        {'basin': 'WP', 'category': [1, 2, 3, 4, 5],
-         'year': 2100, 'change': 1.055, 'variable': 'intensity'},
-        ]
+    gmst_info.update({'rcps' : ['2.6', '4.5', '6.0', '8.5']})
+    gmst_info.update({'gmst_start_year' : 1880})
+    gmst_info.update({'gmst_end_year' : 2100})
```
This is fine. But please initialize the dictionary with one instruction:
```python
gmst_info = {"rcps": ..., "gmst_start_year": ..., ...}
```

peanutfun commented on climada/hazard/test/test_tc_cc.py:
```diff
@@ -26,40 +26,51 @@
-                               0.341930203294547)
-        self.assertAlmostEqual(tc_cc.calc_scale_knutson(ref_year=2078, rcp_scenario=26),
-                               0.312383928930456)
+        criterion = tc_cc.get_knutson_scaling_factor(
+                    percentile='5/10',
+                    baseline=(1950, 2018))
+        self.assertEqual(criterion.shape, (21, 4))
+
+        self.assertEqual(criterion.columns[0], '2.6')
+        self.assertEqual(criterion.columns[1], '4.5')
+        self.assertEqual(criterion.columns[2], '6.0')
+        self.assertEqual(criterion.columns[3], '8.5')
+
+        self.assertAlmostEqual(criterion.loc[2030, '2.6'], -16.13547, 4)
+        self.assertAlmostEqual(criterion.loc[2050, '4.5'], -25.19448, 4)
+        self.assertAlmostEqual(criterion.loc[2070, '6.0'], -31.06633, 4)
+        self.assertAlmostEqual(criterion.loc[2100, '8.5'], -58.98637, 4)
```
If you have to test for non-trivial values here, you need to explain where they come from and how to look them up. Think of it: What is somebody supposed to do once this test suddenly fails? How can one verify which result is "correct"?

peanutfun commented:
@aleeciu I guess this boils down to the issue that we all have limited resources. While we very much welcome the additions in this and other PRs, we cannot provide the same support as for works within our group (which immediately affect our research). We are not used to this situation (yet) and therefore still have trouble managing it.

However, we provide ample guidelines for contributions. In particular, we request contributors to follow PEP8, usual guidelines for unit testing, and our contribution checklist. Particular issues with these can be discussed, but it is the PR author's responsibility to adhere to them as much as possible. As maintainers, we are not responsible to improve PRs ourselves and we reserve the right to close PRs that deviate too much from our (new) standards.

That being said, I think we are almost there. I think we can merge once these points are resolved:
* Please resolve the current discussions. @chahank Please indicate which points need further discussion before merging (in particular with respect to testing).
* Please fix all new linter issues, except `too-many-locals` and `too-many-arguments`.
* > possibly, the notebook I posted above could potentially be simplfied and become and TC_CC tutorial, which is currently missing
   
    Yes, please! I would very much welcome this addition. From a quick glance, this notebook looks nice.

aleeciu commented:
I now restrctured _gmst_ and _knutson_data_ following the inputs from @peanutfun, added better ref to data in the docstring and removed trailing whitespaces. Now I get a petals compatibility error whose meaning I ignore.

I will work on tests in the next weeks. Just for claritiy, these are the guidelines I found: https://climada-python.readthedocs.io/en/latest/guide/Guide_Testing.html#test-files or am I missing more detailed instructions on _how_ to write tests properly?

As for :
> possibly, the notebook I posted above could potentially be simplfied and become and TC_CC tutorial, which is currently missing
> 
> Yes, please! I would very much welcome this addition. From a quick glance, this notebook looks nice.

this is ideal, I will do my best, although tutorials are more about _how_ to use the module and this notebook is more about explaining / validating the methodology behind. I would not consider this a stringent requirement for merging though, it can always be added later relatively easily.

emanuel-schmid commented:
👍 
guidelines: yes, this is the page, including, in particular, https://climada-python.readthedocs.io/en/latest/guide/Guide_Testing.html#unit-tests
plus a rule that seems yet to be missing there but was discussed during climada-days and mentioned in developer meetings: "objects under unit-tests are ought to be created from scratch and not read from files."

chahank commented:
@emanuel-schmid : We should then add this line `"objects under unit-tests ought to be created from scratch and not read from files."` :D 

> this is ideal, I will do my best, although tutorials are more about how to use the module and this notebook is more about explaining / validating the methodology behind. I would not consider this a stringent requirement for merging though, it can always be added later relatively easily.

Fully agree! Thanks for the great improvements @aleeciu. It would be very good to have a description of the methodology somewhere, but I also understand tutorials more as a guide on how to use the module.

aleeciu commented:
Hi, I updated all tests following the guidelines as well as the advise of testing "meaningfull" data and not just prodiving the hardcoded value. I also used np.testing as much as I could following the advide of @peanutfun. From my side, I think I covered all raised points, please let me know.

aleeciu commented:
The only instances where I still check for hardcoded values occurs when testing `get_knutson_data` and `get_gmst_info` whose function is to load data. Based on the previous discussion, having proper tests for these functions was not considered as needed, but I still decided to keep them. They can be removed if needed.

chahank reviewed (changes requested):
Thanks for the improvements! 

Some small typos were found. 

Also, the tutorial file `TropCyclone.ipynb` shows 5301 lines of code changed. This is very suboptimal. Please make sure that only those lines that you changed are changed in this file.

chahank commented:
Oh, and please update the CHANGELOG, don't hide your great work!

https://github.com/CLIMADA-project/climada_python/blob/develop/CHANGELOG.md

spjuhel commented on climada/hazard/trop_cyclone/trop_cyclone.py:
```diff
@@ -366,46 +366,99 @@ def from_tracks(
+
+        tc_cc = copy.deepcopy(self)
+
+        sel_cat05 = np.isin(tc_cc.category, [0, 1, 2, 3, 4, 5])
+        sel_cat03 = np.isin(tc_cc.category, [0, 1, 2, 3])
+        sel_cat45 = np.isin(tc_cc.category, [4, 5])
+
```
Hello, I started using this improvement/fix for the generation of future TCs and discovered that if `TropCyclone.category == []`, this method does not complain at all and just return the exact same object. 

This happened to me when I wrote/read hdf5 files using a mix of `Hazard.write|from_hdf5` and `TropCyclone.write|from_hdf5` instead of only `TropCyclone.write|from_hdf5`, but this could probably happen in other case and is easily addressed by raising an Error if categories are not set, or at the very least a warning, to inform the user.

Here a suggestion (which can probably be improved):

```suggestion
        sel_cat05 = np.isin(tc_cc.category, [0, 1, 2, 3, 4, 5])
        sel_cat03 = np.isin(tc_cc.category, [0, 1, 2, 3])
        sel_cat45 = np.isin(tc_cc.category, [4, 5])

        if sel_cat05.size + sel_cat03.size + sel_cat45.size == 0:
            raise ValueError("Cyclone categories are missing, cannot apply climate change scenario")
```

aleeciu replied:
Thanks for spotting and also for having started using it. I decided to go for the warning.

aleeciu commented on climada/hazard/test/test_tc_cc.py:
```diff
@@ -21,45 +21,127 @@
+            ind2 = target_year_ind + smoothing + 1
+
+            prediction = np.mean(tc_properties[:, ind1:ind2], 1)
+            predicted_change = ((prediction - baseline) / baseline) * 100
```
updated with (hopefully) more informative names: `target_predicted_changes` and `calculated_predicted_change`

aleeciu commented on climada/hazard/trop_cyclone/trop_cyclone.py:
```diff
@@ -366,46 +366,99 @@ def from_tracks(
+        (https://doi.org/10.1007/s00477-021-02142-6). Applying only frequency anyway
+        changes mean intensities and most importantly avoids possible inconsistencies
+        (including possible counting) that may arise from the application of both
+        frequency and intensity changes, as the relatioship between these two is non
```
updated

aleeciu commented:
> Thanks for the improvements!
> 
> Some small typos were found.
> 
> Also, the tutorial file `TropCyclone.ipynb` shows 5301 lines of code changed. This is very suboptimal. Please make sure that only those lines that you changed are changed in this file.

Thanks a lot @chahank . I fixed the typos and the point raised by @spjuhel . Regarding the notebook, I don't know what's going on. I literally changed only one line. So even though it seems huge, it's literally almost the same as before.

I would think we can now merge this one?

aleeciu commented:
It seems that changes to the changelog file were already pushed to develop. It was probably my fault. I can revert it if needed.

chahank commented:
> Regarding the notebook, I don't know what's going on. I literally changed only one line. So even though it seems huge, it's literally almost the same as before.

Thanks for all the updates. As you can see in the files changes, the notebook has 5,056 additions, 245 deletions. Please make sure that only the lines that need to be changed are changed and make a new commit with the correct changes in the notebook.

aleeciu commented:
I have now updated the tutorial withouth re-running it, which seems to have done the trick.

bguillod commented:
I have solved the merge conflicts. For me it's ready to merge, and I think the changes requested by @chahank and @peanutfun have been addressed - can you please confirm by approving and/or by directly merging so we can close this one?
